service_container/service_subscribers.rst

```diff
@@ -0,0 +1,403 @@
+-----------------------------
+
+First, turn ``AbstractUpdater`` into an implementation of
+``ServiceSubscriberInterface`` by adding the static method
+``getSubscribedServices`` which maintains a list of subscribed services and
+replacing our dependencies with a service locator::
+
+    // src/AppBundle/Updates/AbstractUpdater.php
+    namespace AppBundle\Updates;
+
+    use Doctrine\Common\Persistence\ManagerRegistry;
+    use Psr\Container\ContainerInterface;
+    use Symfony\Component\DependencyInjection\ServiceSubscriberInterface;
+    use Twig\Environment;
+
+    abstract class AbstractUpdater implements ServiceSubscriberInterface
+    {
+        protected $locator;
+
+        public function __construct(ContainerInterface $locator)
+        {
+            $this->locator = $locator;
+        }
+
+        public static function getSubscribedServices()
+        {
+            return [
+                'doctrine' => '?'.ManagerRegistry::class,
+                'mailer' => \Swift_Mailer::class,
+                'twig' => Environment::class
+            ];
+        }
+
+
+        abstract public function update($entity);
+
+        /**
+         * Renders a view with Twig
+         */
+        public function render($template, $parameters)
+        {
+            return $this->locator->get('twig')->render($template, $parameters);
+        }
+
+        /**
+         * Sends an email with Swiftmailer
+         */
+        public function send($recipient, $title, $body)
+        {
+            $message = new \Swift_Message();
+
+            // ...
+
+            $this->locator->get('mailer')->send($message);
+        }
+    }
+
+
+With this newly created service subscriber, the updaters can be adapted to
+coincide with the service locator::
+
+    // src/AppBundle/Updates/FooUpdater.php
+    namespace AppBundle\Updates;
+
+    class FooUpdater extends AbstractUpdater
+    {
+        public function update($entity)
+        {
+            // ...
+
+            $body = $this->render('Emails/foo_update.html.twig', [/* ... */]);
+            $this->send($entity->getRecipient(), 'Foo update', $body);
+        }
+    }
+
+    // src/AppBundle/Updates/BarUpdater.php
+    namespace AppBundle\Updates;
+
+    use AppBundle\BarManager;
+    use Psr\Container\ContainerInterface;
+
+    class BarUpdater extends AbstractUpdater
+    {
+        private $barManager;
+
+        public function __construct(ContainerInterface $locator, BarManager $barManager)
+        {
+            parent::__construct($locator);
+
+            $this->barManager = $barManager;
+        }
+
+        public function update($entity)
+        {
+            // ...
+
+            $body = $this->render('Emails/bar_update.html.twig', [/* ... */]);
+            $this->send($entity->getRecipient(), 'Bar update', $body);
+        }
+    }
+
+Optionally, you'll need to add the ``container.service_subscriber`` tag to
+the services definitions of the updaters to enable the service subscribers.
```

Comment:
Thanks to `autoconfigure`, most users won't need this. We usually handle this with just a small note. For example, from https://github.com/symfony/symfony-docs/blob/master/event_dispatcher.rst

> If your methods are not called when an exception is thrown, double-check that you're :ref:`loading services <service-container-services-load-example>` from the EventSubscriber directory and have :ref:`autoconfigure <services-autoconfigure>` enabled. You can also manually add the kernel.event_subscriber tag.